### ISSUE TITLE
Init Templates / Recipes / UI Framework Choice

### ISSUE THREAD

**Issue by jbolda** [tauri-apps/tauri#920]
We want to do our best to remove any barriers for someone trying out Tauri. We have seen new users look to the examples repo which we have maintained thus far as a core testing playground. It has been discussed primarily on discord that we provide a command that makes testing and creating toy apps a breeze.

We have been talking about "recipes" for a bit now. We are thinking we "curry" a command like `create-react-app` and then follow up with the necessary config to set up Tauri to work with it. Like an init command for a greenfield tauri project build with X framework.

An ideal implementation would pull from a list of "templates" which list the command necessary to run and the required config in Tauri that needs to be set. In this manner, we should be able to offload most of the heavy lifting. I think it is open for discussion if this should be part of the init command or it's own separate command. I am imagining something like `tauri init --template` and this enables one of the questions to prompt which framework you would like to use.

Prompts are important so a user isn't expected to type in a framework correctly. It seems pulling from a list of templates is also important as it will enable more easily expanding that if the only requirement is a `.json` file (or equivalent).

I would love input on what others imagine this should be and how we might avoid pitfalls. My expectation is this is scoped to tauri.js and js only libs, and we can add similar as other CLI and bindings are added.

**Comment by zakpatterson** [comment]
My goal for a first PR would be to allow a user to init a new project with a react dev server generated by `create-react-app`, so that it's clear how to add support for other dev frontend choices.

So far, the help and init procedure are:

```
⟩ yarn tauri init -h
yarn run v1.22.4
$ /home/projects/example/node_modules/.bin/tauri init -h
[tauri]: running init

  Description
    Inits the Tauri template. If Tauri cannot find the tauri.conf.json
    it will create one.
  Usage
    $ tauri init
  Options
    [......]
    --plugin, -p         Add UI framework. Supported: react-js
    
Done in 0.26s.

```

* [X] Set up `yarn tauri init` CLI prompt changes
* [X] Recipes can specify NPM dev and prod dependencies
* [x] Recipes can specify post-installation instructions, such as `create-react-app`
* [ ] Identify and update documentation via PR:
  * https://tauri.studio/en/docs/development/
  * https://tauri.studio/en/docs/development/integration#2-initialize-tauri

```
⟩ yarn tauri init
yarn run v1.22.4
$ /home/projects/example/node_modules/.bin/tauri init
[tauri]: running init
? What is your app name? 
? What should the window title be? Tauri App
? Would you like to add a UI plugin? (Use arrow keys)
❯ None 
  react-js 
```

No plugin selected by default. Note also that the choice of dev server URI would be controlled by plugin setting. None => Prompt for dev server URI, any other selection would assume that the devPath would be set by the plugin settings.

**Comment by nothingismagick** [comment]
This is cool, a nice start. But the word plugin means something else. Would you please consider using 'recipe'?

**Comment by martonlanga** [comment]
Next.js calls them `examples`. `Boilerplate` & `recipe` is also nice.

**Comment by jbolda** [comment]
One thing that @zakpatterson and I were discussing is the folder structure. We don't hold any particular strong opinions regarding folder structure and tauri (`init` does use `src-tauri` which may be the only one?), but we will need an opinion for this. I have generally assumed my root held the main package.json, an `src` for the ui/web frontend and `src-tauri`. Most of these init scripts (like `create-react-app` for instance) will create a folder for you to my knowledge. To support this structure, we would likely need to run `create-react-app` first (which creates the folder for you), and then lay tauri on top.

The current iteration of #930 assumes multiple package.json. The user creates the folder, adds a package.json (because tauri requires it), inits tauri and then runs `create-react-app`. This gives us a folder structure like the following. I don't know that either structure is "wrong" per se, but one vs the other kind of dictates the order we run these and the tauri assumptions as to what exists when running init.

![image](https://user-images.githubusercontent.com/2019387/89783211-2e241b80-dadc-11ea-8f4e-513fe4b1bf8d.png)

Thoughts? Are there any frameworks out there that init differently than `create-react-app` that we need to consider?
/cc @tauri-apps/core

**Comment by jbolda** [comment]
Should this be a top level command? Along the lines of `tauri recipe` that then hooks into the init sequence after choosing `create-react-app` or whatever. I think `tauri init` feels moreso intended towards bootstrapping tauri into an existing project. Additionally, we could support `tauri recipe` which runs your UI init script and then layers on tauri config, and also have `tauri init` and choosing a "recipe" assuming you have already run `create-react-app` and only takes care of the tauri config. Essentially `tauri recipe` for a greenfield project, `tauri init` for a brownfield project (but use some of the effort from `tauri recipe` to preset config).

**Comment by nothingismagick** [comment]
That's a great idea really @jbolda

**Comment by zakpatterson** [comment]
Sure we could put the UI project source, configuration, and build output in the same root directory as the main project. If the UI project is in a "js" framework, then it could share the same package.json file as the root project.

But then if the UI project were written in something else, like Scala,js, then the root directory would get a `build.sbt` file. Scala also puts its build products into a folder called `target`, so something else would need to be done to avoid colliding with rust's target folder.

For the sake of generality, it makes more sense to me at least for the UI code, build conf, and build output to go into one subdirectory by default. I get that's not how the examples are done now. Otherwise, I agree, I think we'd have to try the recipe first and then `init`, happy to do it that way if it's preferred.

Also I started on #930 without looking through the PRs and roadmap, and I never saw #851. Should I put more effort on advancing #851 first?

**Comment by jbolda** [comment]
The more I think about it, the more I tend agree. It probably makes sense to keep it consistent across the board. The being said, it probably makes sense to push towards using yarn for the workspaces. We need a root package.json for tauri, but most of these starter CLIs also add their own package.json. We could normalize everything from the root then I imagine.

Regarding #851, I think that is kind of waiting on #850 to reconcile things around that. We expect to always have a js CLI, but plan to add CLIs for other languages as well. Moving some of the core logic into rust means that we can share that between all of the CLIs. I don't know how clear that divide is, but I imagine the "recipe" logic makes sense to be implemented generally in each language. Then each can deal with the specfic language nuances there.

/cc @lucasfernog @tensor-programming ^ make sense to you?

**Comment by zakpatterson** [comment]
I'm not sure I understand the use case for `init` to just add Tauri to an existing project. Should I add another top level command for `recipe` that only modifies an existing build configuration and runs the ui init script?

**Comment by nklayman** [comment]
First of all, I like the idea of the PR and how it can make it much easier to get started with Tauri. I do have a few suggestions though:

First, I think that the CRA app files should be at the root level, with the `src-tauri` folder inside of it. It's how all the examples and the Vue CLI plugin are structured. The issue with Scala using `target` for dist isn't really a problem since Tauri dist is in `src-tauri/target` (although I do see how it could cause some confusion). Having a different structure than an existing CRA app with Tauri added via `tauri init` seems like unnecessary fragmentation. I think 95-99% of users will be using a JS framework and have a `package.json` anyway and sharing it would be easier to manage. Lots of new developers may try to install deps for their JS app into the tauri `package.json` and would be confused when it doesn't work.

Second, maybe we could automatically set [beforeDevCommand](https://tauri.studio/en/docs/api/config#build.beforeDevCommand) to start the CRA dev server so that `yarn tauri dev` just works, instead of making the user run multiple commands (same for build).

Third, I think there should be two separate commands for using a recipe and adding to an existing app (like @jbolda suggested). Maybe `tauri create` for creating a new app with a recipe, and `tauri integrate` (or `tauri init`) for adding to an existing app. I think this will make the process easier to understand for new developers trying to decide how to get started with Tauri.

Finally, I'm a little worried that this may be out of scope for the Tauri CLI. Creating a full list of recipes and keeping them updated isn't an easy task. We'd have to monitor for breaking changes in each CLI we use and then update Tauri CLI to support the latest version. What if a user wants to pass arguments to their framework CLI's init command? Do we have to create a passthrough for all the different config args for each framework? We also want to migrate the CLI to Rust at some point, and this would make that more difficult. A getting started doc page with instructions for each of the frameworks would be easier to maintain and give the user more control, although it would sacrifice some ease-of-use.

**Comment by zakpatterson** [comment]
The scope concerns are totally understandable. But it seems people agree something must be done because it's a common speed-bump for beginners. 

Maybe `init`/`create`/`recipe` doesn't even belong in this repository? Maybe we need  `create-tauri-app` as a separate tool.

**Comment by nklayman** [comment]
I actually really like the idea of `create-tauri-app`. It will make the Rust CLI transition easier since it could stay as JS. It would still be really easy for the user as they could just run `yarn create tauri-app` and then they're good to go.

**Comment by jbolda** [comment]
@zakpatterson the thought with `init` in an existing project, I think is the expectation that has a web app and wants to add desktop app. That is currently the only init, but I imagine it would see less use with a recipe option as I expect the majority are greenfield projects.

I don't think we expected to have a package.json in `src-tauri`, @nklayman. (At least I didn't.) The root could operate off of yarn workspace, but that is extra to document and gets away from the value of recipes. I think whichever structure we choose will set a precedent. I don't know that we need to lean heavily on historical context being <1.0. The bigger question is in the context of future languages and bindings I think.

I agree with the beforeDevCommand, but I don't know how that would work with yarn workspaces (haven't actually used it yet myself).

My understanding is that there will still be a thin wrapper of js that I think this would new fine to fit within. I mentioned this earlier, but I don't think it makes sense for js to handle every languages "recipes" nor Rust necessarily either. I think that leads to some precedent for `create-tauri-app`, but other languages don't really do that (but maybe it doesn't matter but consistency is also nice).

At the end of the day though, we say this works with any UI framework so we will need some indication or documentation for every one otherwise it will come across as an empty promise. If this implementation only concerns itself tauri config and pass-through another command, the effort required is really low and will make getting into tauri a breeze. This brings in more people trying it out and seeing the possibilities, but also makes it much easier to spin up and test different frameworks. Marketing is a lot harder then code and even docs.

Where we place it is probably more open ended in my mind. I don't think we ever decided what the Rust CLI core would actually include?

**Comment by zakpatterson** [comment]
Ok, parsing the above from everyone, thank you for feedback, I suggest this:

* [x] [New top level command `tauri create`](https://github.com/tauri-apps/tauri/pull/930/commits/b4ae4df45edf63e5ba49cc25e790486742ce2cf0), which is basically the same as `tauri init` except that it can use a recipe to build the UI framework. Just so we're clear I'm going to implement this by renaming the current `init` to `create` with the extra `-r` flag, and creating a new command `init` which is an alias to `create -r none`.
* [x] No functional change to current `tauri init`, [aliased to `create`](https://github.com/tauri-apps/tauri/pull/930/commits/99d491222916fa6f9d431a1bf9e05dc0ecb47977).
* [x] [React recipes set `beforeDevCommand` and `beforeBuildCommand`](https://github.com/tauri-apps/tauri/pull/930/commits/83f3576514b857258d4a692b066a3a657d8ca184)
* [ ] I still think the UI build configuration stuff should be separate from the tauri build, the `beforeDevCommand` and `beforeBuildCommand` settings make this much nicer. 
   The point about installing packages to the web ui front-end only is a great one. But adding packages to the rust side of the desktop app "middle-end" behaves the same way. The `Cargo.toml` file is inside of the src-tauri folder. It just seems more congruent to me if the UI side of the desktop app has a parallel directory structure to the rust side.
* [ ] Open new issue for `create-tauri-app`.

**Comment by chippers** [comment]
> Regarding #851, I think that is kind of waiting on #850 to reconcile things around that.

#850 doesn't touch the cli at all besides changing the template, so I'm not aware if it's waiting for it beforehand

edit: oh this is closed lol
